spring-grpc-spring-boot-autoconfigure/src/main/java/org/springframework/grpc/autoconfigure/server/GrpcServerAutoConfiguration.java

```diff
@@ -17,36 +17,51 @@
 
 import java.util.List;
 
-import org.springframework.boot.autoconfigure.AutoConfigureBefore;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import io.grpc.BindableService;
+
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.AutoConfiguration;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
-import org.springframework.context.annotation.Configuration;
 import org.springframework.grpc.server.DefaultGrpcServerFactory;
 import org.springframework.grpc.server.GrpcServerConfigurer;
 import org.springframework.grpc.server.GrpcServerFactory;
+import org.springframework.grpc.server.lifecycle.GrpcServerLifecycle;
 
-import io.grpc.BindableService;
-import io.grpc.ServerServiceDefinition;
-
-@Configuration(proxyBeanMethods = false)
-@ConditionalOnMissingBean(GrpcServerFactory.class)
-@ConditionalOnBean(BindableService.class)
-@AutoConfigureBefore(GrpcServiceAutoConfiguration.class)
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for gRPC server-side components.
+ *
+ * @author David Syer
+ * @author Chris Bono
+ */
+@AutoConfiguration
+@ConditionalOnClass(BindableService.class)
```

Comment:
I believe this was a bug before as `@ConditionalOnBean` would not fire since this is the config that adds the BS beans. Also, we want to disable if they do not have `io.grpc` on the classpath.

Reply:
> and the integration tests seem to prove it)

Ahh, I just noticed these under `samples` module. 👍🏻 

> The BS beans are added by marking them as `@Component`

I missed this fact. So basically, if there are no BS beans then there is no need to auto-config the server. Would it ever be the case that one would want to add a service after startup? If so, we may want to remove this constraint. If we did, then the object provider would just return zero services and the server would start but not provide any services at that point.

> Why would you be using spring-grpc without io.grpc on the classpath?

The scenario is when this later lives in Spring Boot AC module it will be necesarry.
I am approaching the auto-configuration (AC) as if it lives in Spring Boot because:

1. I expect that is where we want it long term
2. the more it looks like that then it will be easier to make the move

I did this for spring-pulsar and it made the move into SB straightforward.

Reply:
> and the integration tests seem to prove it

I believe the ITs are passing.

Reply:
That's not the point. The point is that they were passing before, so you can put the `@ConditionalOnBean` back in.

Reply:
Done. I also added an auto-configuration test for the server config.